### ISSUE TITLE
Almost all Face Landmarks are (undesirably) affected by eyes and irises motion

### ISSUE THREAD

**Issue by ianni67** [google/mediapipe#1786]
Hello, Thank you for sharing this great piece of software. 
I'm writing here to evidence an undesirable behavior of the face landmarks module that you might want to correct.
I'm working on eye tracking and noted that most of the face landmarks move every time the irises move. This leads to wrong variations in the face pose transformation matrix, which, in turn, _affects most applications of the face geometry module_.
This behavior can be easily demonstrated and was also reported in this [comment](https://github.com/google/mediapipe/issues/1642#issuecomment-804362081).
I'm currently looking for a workaround for my specific case, but I'm unlikely to succeed until a general solution is found.

[EDIT:] Just to be more clear, it would be great if the face pose transformation matrix would not depend on iris and (small) eye motion.

**Comment by kartynnik** [comment]
1. This has been to an extent addressed by the [Attention Mesh](https://arxiv.org/abs/2006.10962) model. There are (yet unspecified) plans to release a light attention mesh model in the future.
2. One might try to decrease the effect of the undesirable movement of some landmarks by adjusting their weights in the canonical face mesh fitting procedure. See the lines like `procrustes_landmark_basis { landmark_id: 4 weight: 0.070909939706326 }` in `geometry_pipeline_metadata_landmarks.pbtxt` and the canonical face model in `canonical_model.*` under https://github.com/google/mediapipe/tree/master/mediapipe/modules/face_geometry/data.

Let @kostyaby comment on anything I might be missing.

**Comment by ianni67** [comment]
@kartynnik thank you very much for your answer, which I was eagerly awaiting. Just to check if I'm getting it right, your current suggestion is to decrease the weights of the eyes-related landmarks in the canonical face mesh fitting procedure, in order to obtain a transformation matrix that is less sensitive to eyes and irises motion. Did I understood correctly?
The file you pointed out (geometry_pipeline_metadata_landmarks.pbtxt) only contains a few weights. How can I adjust, for example, the weights of landmarks 159 and 160, that are not included in the file?

**Comment by kartynnik** [comment]
@ianni67 As can be seen [here](https://github.com/google/mediapipe/blob/350fbb2100ad531bc110b93aaea23d96af5a5064/mediapipe/modules/face_geometry/libs/geometry_pipeline.cc#L446), the other weights are set to zero.

**Comment by ianni67** [comment]
So, those weights are currently the only ones that actually affect the fitting?

**Comment by kartynnik** [comment]
You're correct. You might actually observe more stability if you add more points participating in the fitting rather than removing them. See the [Kanstantsin's post](https://developers.googleblog.com/2020/09/mediapipe-3d-face-transform.html) for the background on what that fitting entails.

@kostyaby Can you please share how the weights were assigned?

**Comment by kostyaby** [comment]
Hey @ianni67,

I'm actually less optimistic than @kartynnik on the matter of how much the face geometry fitting pipeline weight tweaking can help to achieve your goals. Although [these GIFs](https://github.com/google/mediapipe/issues/1642#issuecomment-804362081) might make you think that there only a small landmark motion that causes a huge face pose motion, I'd say this opinion doesn't take into account the following visual illusion: to a human eye, one large rigid object motion seems much more severe than little movements of individual landmarks. My opinion is that such head transform movement is within expected range given such individual landmark movement. Therefore, I’m a bit reluctant to suggest you playing with weights as I don’t see any significant gains you might reach there. It could make things a little bit better, but probably not in a major way

As for the weight assigning process, I'd call it empirical "trial-and-error" style experimentation. There were a couple of guiding principles that I was was following while generating the weights (I'll list them below), however I cannot provide you with a clear guidance on how to replicate / adjust this process. The guiding principles where:

- Naturally, some landmarks are much more receptive to "intrinsic" face motions (for example, facial expressions); such landmarks should be excluded from the fitting process (or, if completely necessary, assigned a very-very low weight) as we are interested in getting the "extrinsic" face motion (=face pose);
- Symmetric landmarks (for example, the left corner of the left eye and the right corner of the right eye) should have the same weight;
- The center of mass for the weighted fitting landmark basis should be very close to the center of mass for the entire face landmark set of 468 points; this leads to a better rotation / scale estimation during fitting as the Procrustes solver depends on the center of mass to remove the translation part first;
- And finally (this principle is the most informal and thus the main source of the empirical nature) - when applied to drive an AR asset, its motion should "feel natural", as if it was a real item attached to one's head. I had iterations when I had fitting pipeline candidates which were better in terms of quantitative metrics I had, but then resulted into a fairly unnatural motion when used with AR elements - therefore, I couldn't give such candidates a go as my quantitative metrics were clearly not good enough to blindly guide the algorithm selection process and it was unclear how to put the "natural motion feel" into a metric to track

In [this Github issue](https://github.com/google/mediapipe/issues/1854#issue-852226689), another user visualized the current weights present in `geometry_pipeline_metadata_landmarks.pbtxt`. If you feel like trying to play with weights, this would be a good starting point to locate landmark around eyes from the fitting basis & try decreasing their weight; it _might_ help you use-case. However, if you won't be able to make it right within a few hours / days, I wouldn't invest weeks of work into tweaking the weights. As one can clearly see on [these GIFs](https://github.com/google/mediapipe/issues/1642#issuecomment-804362081), face mesh landmarks clearly react to eye motion and ideally they shouldn't react at all. Fixing the face mesh NN model is a better, fundamental solution; after that's done, Face Geometry fitting pipeline will work like charm with the current configuration

**Comment by ianni67** [comment]
Hello @kostyaby and @kartynnik.
First of all, let me thank you again for your kind availability and clear explanations. 
I believe I got all your arguments. I also think that the main point here is that my target application is substantially different from that of "putting a mask" on the face of a subject. The intuitive reason is that when the objective is sticking a mask to the face, the landmarks should actually move also according to the facial muscles and not only with the skull movements. In other words, if we barely let the mask stick to the skull, it may look "unrealistic" and somewhat unnatural in some cases. Masks _should_ stick to the skin and move accordingly to facial muscles, as in this way they fit better. Similarly, when I look at my face at the mirror and move my eyes up/down, my glasses actually move a bit. This happens because the actual visual range for "vertical" eyeball rotations is smaller than that for lateral ("horizontal") rotations and, thus, a motion of some facial muscles is involved when "extreme" vertical rotations are made.
My conclusion is that the "defects" of the current modules can somehow be considered as acceptable for mask fitting, whereas they are not acceptable for my application, that requires a reliable face rotation matrix that actually reflects "bones" orientation and rotations. 
I will probably keep on looking for alternative solutions for my specific case. When (and if) a general solution will be released, I will be happy to adopt it. I will keep you informed about my (eventual) progress.

**Comment by lilinxiong** [comment]
Hi @ianni67 I think what we want to achieve is the same function.
I need too "rotation matrix that actually reflects "bones" orientation and rotations."
If an alternative is found. Can you share it here ?
The same, if I find a good solution, I will share it here～
Thanks~

**Comment by ianni67** [comment]
I'm most probably not going to find a general solution for the problem. My specific objective might, however, have an empiric, perhaps ad-hoc solution that I will post here (when I find it).

**Comment by lilinxiong** [comment]
Looking forward to your plan.

**Comment by lilinxiong** [comment]
Hi @ianni67 
I have been looking for a reliable face rotation matrix during this period of time, through the official documentation and search from the code.
Then I got the very accurate Euler angle parameters I wanted the day before yesterday, because getting Euler angles is what I need in my current work. Of course, I have to get accurate Euler angles from a reliable rotation matrix.
In short, I got an accurate rotation matrix...
But I’m not sure if our goals are the same.
I recorded my effect into a video, if you are interested, I can send it to your email...

**Comment by ianni67** [comment]
> Hi @ianni67
> I have been looking for a reliable face rotation matrix during this period of time, through the official documentation and search from the code.
> Then I got the very accurate Euler angle parameters I wanted the day before yesterday, because getting Euler angles is what I need in my current work. Of course, I have to get accurate Euler angles from a reliable rotation matrix.
> In short, I got an accurate rotation matrix...
> But I’m not sure if our goals are the same.
> I recorded my effect into a video, if you are interested, I can send it to your email...

I'm definitely interested, so please if you can, send me the video.
I've also been working on a solution on my side. I realized that some of the landmarks are sufficiently stable and I'm currently relying on those landmarks only for my application. The solution is specific to my application but seems to work.
If you are interested I can let you know how it works.

**Comment by lilinxiong** [comment]
I will send to ianni67@yahoo.com.

**Comment by ianni67** [comment]
Just as a follow-up, I'd like to share a couple findings.
- Some landmarks indeed seem sufficiently stable. For example landmarks 02 and 164 appear to be fairly independent of the irises motion. I made a few tests and decided to rely on those reasonably stable landmarks for my application. I suppose that by changing (indeed, adding) their weights in geometry_pipeline_metadata_landmarks.pbtxt, while also setting to a very low value (virtually zero) the weights of the landmarks near to the eyes, might lead to better results for those specific cases that need a model that "sticks to the skull". However, I did not find the time to test this approach.
-  As a side-effect, we (together with @lilinxiong) found that the eyebrows landmarks do not fit correctly to the eyebrows of the subject. We both made experiments and both had similar results. The eyebrows landmarks are often very unreliable.
This can be easily demonstrated by rotating up-down the head and observing the eyebrows landmarks.
Having unreliable eyebrows may lead to issues for applications related to emotion detection.

**Comment by lilinxiong** [comment]
Yeah yes, I very much agree @ianni67 said.
Let me add one more point, I only noticed yesterday.
* In FaceMesh solution. I find the eyes landmarks never change,no matter if you open your eyes or close your eyes, initially I suspected that my configuration was wrong. But when I saw [face-landmark-model-Fig 2. Face landmarks](https://google.github.io/mediapipe/solutions/face_mesh#face-landmark-model) gif again, the eyes landmarks inside will not change. So is this a problem with the model itself ? 
  Then I have a question in [FaceMesh can be adopted to drive virtual avatars (middle).](https://ai.googleblog.com/2020/08/mediapipe-iris-real-time-iris-tracking.html) in this demo effect, first FaceMesh eyes landmarks never change(open eyes and close eyes), but can blink, this is incredible.

**Comment by Eunchan24** [comment]
> Hi @ianni67
> I have been looking for a reliable face rotation matrix during this period of time, through the official documentation and search from the code.
> Then I got the very accurate Euler angle parameters I wanted the day before yesterday, because getting Euler angles is what I need in my current work. Of course, I have to get accurate Euler angles from a reliable rotation matrix.
> In short, I got an accurate rotation matrix...
> But I’m not sure if our goals are the same.
> I recorded my effect into a video, if you are interested, I can send it to your email...

Hello, I am also interested in your video. Can you send it to me?

**Comment by ianni67** [comment]
Hello everyone, I thought that probably the issue has been fixed?
Are facial features more stable now, with respect to eyeballs motion? I'd love to resume my original project and complete my proof-of-concept.
Thank you very much for your help!

**Comment by kostyaby** [comment]
Hey @ianni67,

The recent MediaPipe release ([v0.8.8](https://github.com/google/mediapipe/releases/tag/v0.8.8)) has the [Attention Mesh](https://google.github.io/mediapipe/solutions/face_mesh#attention-mesh-model). It definitely has much better metrics around eye / lip landmarks, however I'm not 100% sure it has a better quality for your particular use-case. Maybe you can give it a shot?

**Comment by ianni67** [comment]
Hey @kostyaby,
thank you for answering! Indeed I gave Attention Mesh results a shot. As far as I could get, the problem is still there: the accuracy of eyes and lips landmarks has indeed been improved, but the other landmarks are still affected by the motion of irises, so my application still does not work :-(

**Comment by kostyaby** [comment]
I see, thanks for checking out. It's too bad the quality is still not enough; AFAIK this is close to our best tech available, so if it doesn't work then I'd say there's not a big chance anything based on MediaPipe face tracking would work (based on the progress we have so far)

**Comment by ianni67** [comment]
It may be a problem of my approach as well: perhaps I'm not exploiting 
MediaPipe in the best way.

My original idea was to track both the iris with respect to the head, 
and the 3d head orientation with respect to some absolute 3D space. The 
output of the process should be an estimate of the eye gaze direction in 
the 3D space, and I also needed some projection matrix allowing me to 
calibrate the eye gaze estimator, so that I can estimate the point where 
the 3D vector of the eye gaze direction hits the screen plane. In such a 
way I should be able to estimate where the user is looking at, on the 
computer screen.

Unfortunately I met a lot of issues, the first being the accurate 
estimation of the head orientation. Then I had troubles in calibrating 
the projection matrix of the 3D head orientation onto the screen space. 
Finally, I reverted to looking for a set of fiducial landmarks on the 
face, that were sufficiently independent from the eyeball motion, to be 
used as a reference for calculating the eye gaze direction with respect 
to the head. But this did not work either, because most landmarks are 
indeed affected by eyeball motion.

I'll look into the whole story, for a way to solve my issues.

Thank you very much for your help

Giancarlo

Il 03/12/21 05:08, Kanstantsin Sokal ha scritto:

**Comment by kostyaby** [comment]
Hey @ianni67,

Thanks for providing your perspective!

I guess that's just the nature of the Computer Vision field: everything *kinda works*, but then you have to solve a particular use-case and it's not clear whether the underlying tech is good enough and just needs more glue logic or is it not fitting the use-case entirely. Sounds like you gave MediaPipe a good try and it doesn't seem to solve your problem - that might be a good indicator to see what other options are available out there. For your particular use-case, maybe having a CNN that'd map eye images directly into screen XY coordinates ([similar work](https://ai.googleblog.com/2021/05/accelerating-eye-movement-research-for.html)) would be a better approach

**Comment by ianni67** [comment]
Hello @*kostyaby, <https://github.com/kostyaby>*

Thank you for pointing me to that paper. It sounds very promising.

I indeed knew that one of the mainstream threads was investigating the 
direct use of DNN to map eye  geometry directly to gaze target. 
Something similar has been described a couple years ago. Of course, at 
that time the presented work was not complete.

As a long-term survivor in the area of computer vision and applied AI, 
I'm also aware that I'm swimming in the same pool with much larger and 
powerful fishes. Nevertheless, I'm convinced that investigating diverse 
approaches might lead me to some interesting, alternative solutions - 
maybe, I'll survive for some more years ;-)

I'll give a try to a few other solution, maybe I'll find the right one.

Thank you very much for your help. I'll keep in contact

ianni

Il 06/12/21 22:00, Kanstantsin Sokal ha scritto:

**Comment by ianni67** [comment]
Hello @kostyaby, I'm sorry for bothering again, but I'm currently looking into the python version of mediapipe and found out that, according to https://github.com/google/mediapipe/issues/1530#issuecomment-940746284,  the iris landmarks are now included within the face mesh model when the refine_landmarks configuration option is set to True. However, I could not find a file where the (3D) coordinates of such landmarks in the canonical face model are available. 
In other words, I do not know where to find the coordinates of the iris landmarks from the new canonical face model. 
Would you, please, help me once again and provide a hint or a pointer towards the right direction?

**Comment by kostyaby** [comment]
Hey @ianni67,

It is true that the canonical face model does not have iris landmarks. It's also true that the iris landmark NN generates 2D landmarks, not 3D landmarks. Thus some sort of iris landmark Z coordinate deconstruction has to be done; face landmarks are in 3D, so it'd make most sense to start exploring from there

Reg. adding iris landmarks to the canonical face model: task has a simple-ish heuristic solution with a decent approximation. The average human interpupillary distance (IPD) is 64mm - that's the distance between the leftmost and the rightmost iris landmarks (please see the [IrisToDepthCalculator](https://github.com/google/mediapipe/blob/ecb5b5f44ab23ea620ef97a479407c699e424aa7/mediapipe/graphs/iris_tracking/calculators/iris_to_depth_calculator.cc)); canonical face model is in centimeters, so you can roughly compute the scale for iris landmarks. Now, those iris landmarks are located approximately on a 3D sphere whose XY center is in between edge eye landmarks and Z center is a bit deeper inside the head (a parameter to play with). For each eye, you should find params for such 3D sphere ("eyeball") and then place iris landmarks semantically according to the "IPD ~ 64mm" scale factor.

Reg. generating Z coord for iris landmarks during runtime, this one is already quite hard. You can still use all the same observations from the previous point (iris landmarks are located on a 3D sphere; IDP ~ 64mm). Maybe, you can convert face landmarks into the Metric AR space first (thus making them roughly in centimeters); from there, you can use the transform matrix to place the canonical 3D "eyeball" spheres into the Metric AR space; then, you know that the 2D iris landmarks should be a perspective projection of some points from the surface of those 3D "eyeball" spheres - so now you have to solve the reverse problem and "unproject" 2D iris landmarks back into the 3D "eyeball" spheres (knowing the perspective camera matrix that projects them into the screen 2D coordinates)

My general concern about iris landmarks is that we are talking about a tiny object (~64mm) of a much larger object (faces can be ~20cm long), so the requirement for precision is much higher - meaning any heuristic that'd yield an otherwise "decent" approximation can now introduce severe artifacts (depending on a use-case)

**Comment by ianni67** [comment]
Hello @kostyaby, 
I agree, the main concern in this topic and for this specific approach is accuracy and stability of the results. I got some decent approximation of eye gaze direction using different approaches but the stability (i.e., the coherency of the measured eye gaze angles with the actual eye gaze direction in a time interval) is generally poor. Moreover, calibration when the camera and face positions are both non-fixed is hard. 
Thank you for your clarifications, I guess those will also help other users in correctly approaching mediapipe development.

**Comment by ianni67** [comment]
> Maybe, you can convert face landmarks into the Metric AR space first (thus making them roughly in centimeters); from there, you can use the transform matrix to place the canonical 3D "eyeball" spheres into the Metric AR space; then, you know that the 2D iris landmarks should be a perspective projection of some points from the surface of those 3D "eyeball" spheres - so now you have to solve the reverse problem and "unproject" 2D iris landmarks back into the 3D "eyeball" spheres (knowing the perspective camera matrix that projects them into the screen 2D coordinates)

Hello @kostyaby, thank you very very much for finding the time to keep this discussion open and providing precious information to those who are developing with mediapipe. 
Wouldn't it be much easier if the code to convert the face landmarks into the Metric AR space would also include the iris landmarks newly added to the "attention model"? Currently the geometry pipeline only processes the first 468 landmarks, while with the new model, 478 landmarks are produced, including iris landmarks. It would suffice to add a few points to the canonical face model and a few changes in the geometry_pipeline calculator.

Note: To be more precise, the geometry pipeline crashed when trying to process the face landmarks produced with the attention model: I had to apply a few changes to the  geometry pipeline calculator, proposed by @lilinxiong in issue #2867. 
As a consequence, allowing the geometry pipeline to process all 478 landmarks would also fix an incoherence issue within mediapipe.

**Comment by kostyaby** [comment]
> Wouldn't it be much easier if the code to convert the face landmarks into the Metric AR space would also include the iris landmarks newly added to the "attention model"? Currently the geometry pipeline only processes the first 468 landmarks, while with the new model, 478 landmarks are produced, including iris landmarks.

In an ideal world, I can see the Face Geometry module having multiple operation modes (face landmark-only; face landmark + iris landmark; maybe some other combination in the future). However, currently the effort is not staffed to cover all possible combinations. At this point, all I've committed to is supporting the existing Face Geometry module based only on the original face landmark NN output. Supporting the attention model NN output is an additional non-trivial feature which is not staffed from our side at this moment

> Note: To be more precise, the geometry pipeline crashed when trying to process the face landmarks produced with the attention model

I recognize this is a bad dev experience and I'm sorry you had to go through this with MediaPipe. Ideally, things should "cling" together effortlessly. Realistically, it's hard to maintain all possible input combinations while allowing the degree of flexibility MediaPipe offers. As you've probably already noticed, MediaPipe is quite research-y and is really about putting great ML/CV building blocks out there and demonstrating *some* combinations of them working together, but it's not always well-designed to plug-n-play with any combination of the building blocks. Should the Face Geometry module support the attention model output? Yeah, that'd be ideal! Considering the lack of resources, should the attention model release be postponed because Face Geometry doesn't support its output? In the current MediaPipe strategy, the answer is "no, don't postpone it, the attention model is quite valuable in itself"

Regarding #2867, Face Geometry module still perfectly supports a 468 landmark list input, so my recommendation would be to use [Split*LandmarkListCalculator](https://github.com/google/mediapipe/blob/139237092fedef164a17840aceb4e628244f8173/mediapipe/calculators/core/split_landmarks_calculator.cc#L162-L168), select the first 468 landmarks out of 478 produced by the attention NN and use that as an input to the `FaceGeometryFromLandmarks` subgraph. If you want to have iris landmarks in the Metric AR space, you'd have to do additional logic like I described in my previous comment. Again, I currently lack necessary eng cycles to do this job myself as there are more high priority work that requires my attention internally at Google. 

I'd refrain from using `Split*LandmarkListCalculator` by default inside `FaceGeometryFromLandmarks` as the abstraction starts to leak: when the input is 478 landmarks and the output is 468 landmarks, it might be confusing - "why not all input landmarks are converted?" users might ask. I believe it's better when `FaceGeometryFromLandmarks` explicitly crashes early when it cannot fulfill the contract and it lets you, as an end-user, to manually and consciously remove 10 iris landmarks. This way, it follows the Principle of Least Astonishment better IMO

I'll be closing this issue as I believe this discussion is no longer about the original question. Feel free to open new issues to discuss things further

**Comment by ianni67** [comment]
> I'll be closing this issue as I believe this discussion is no longer about the original question. Feel free to open new issues to discuss things further

I see. Sorry for that. I guess the discussion drove me too far.
Thank you for taking the time to answer our questions.
